python/heterocl/schedule.py

```diff
@@ -292,66 +286,298 @@ def to(self, tensors, dst, src=None, axis=0,
             Move axis-th loop body to xcel scope
 
         mode : data movement type
-            The modes of data movement (FIFO, DMA, MMIO)
-            For inter-kernel data movemnet, only FIFO is supported
+            The modes of data movement (Stream, DMA, MMIO)
+            For inter-kernel data movemnet, only Stream is supported
 
         depth : channel depth
             The streaming channel depth
-
-        local_buffer : boolean
-            create local buffer for data on-device
-
+            We leave an interface here to specify the FIFO depth
+            in the future we should be able to infer automatically
         """
-        if mode not in [ _expr.IO.DMA, _expr.IO.FIFO ]:
-            raise APIError("Invalid channel type")
+        if mode not in [ _expr.IO.DMA, _expr.IO.Stream ]:
+            raise APIError("Only DMA and Streaming modes are supported...")
+
+        # support chained .to()
+        if dst is None:
+            dst = tensors
+            assert self.hold_tensor is not None
+            tensors = self.hold_tensor
+            # hold stage has none value when the previous move is to device
+            if self.hold_source_stage is not None:
+                src = self.hold_source_stage
+
+        # handle more than one input tensors for data movement 
         rets = list()
         if not isinstance(tensors, list):
             tensors = [tensors]
+
+        # handles many to many
+        if len(tensors) > 1: 
+            if isinstance(dst, list):
+                assert len(dst) == 1
+        if isinstance(dst, list):
+            if len(dst) == 1:
+                assert len(tensors) == 1
+
+        # one-to-many or many-to-one
+        # handle more than one dest (multi-casting)
+        if isinstance(dst, list):
+            for d in dst:
+                self.to(tensors, d, src, axis, mode, depth, burst, burst_len, name)
+            return self
+        
         for tensor in tensors:
             try:
+                # move the output tensor of a stage
                 if isinstance(tensor, Stage):
                     target = tensor._op
+
                 # unpack tuple of src stage and tensor
+                # E.g. kernel.stage.B = (stage, B)
                 elif isinstance(tensor, tuple):
                     src, target = tensor
-                    # from hcl stage to tvm stage
+                    # from heterocl stage to tvm stage
                     src = self.__getitem__(src)
```

Comment:
I changed the interface a bit -- if users access a tensor with `Stage.tensor`, I will let HCL return a tuple of (Stage, Tensor). Since in some cases, we need both these two objects to mutate the IR

Reply:
So when do we use `Stage.tensor`? The user needs to decide?

Reply:
Fo example, we create a tensor in stage A, it's updated in stage B and stage C. The user need to decide when he wants to tensor to be moved back to host

Reply:
```python
s.to(StageB.tensor. p.host)
s.to(StageC.tensor. p.host)
```